README.md
<p align="center">
  <a href="https://melpa.org/#/circadian" target="_blank">
    <img src="https://melpa.org/packages/circadian-badge.svg" alt="MELPA"/>
  </a>
  <br>
  <a href="https://github.com/guidoschmidt/circadian.el/actions/workflows/ci.yml" target="_blank">
    <img src="https://github.com/guidoschmidt/circadian.el/actions/workflows/ci.yml/badge.svg"
         alt="https://travis-ci.org/guidoschmidt/circadian.el"/>
  </a>
  <br>
  <br>
  <br>
  <img src="logo.png" alt="Logo"/>

  <h1 align="center"><a href="https://guidoschmidt.github.io/circadian.el" target="_blank">circadian</a></h1>
  <h3 align="center">Theme-switching for Emacs based on daytime</h3>
</p>

### Conception

Circadian tries to help reducing eye strain that may arise
from difference of your display brightness and the
surrounding light.

Inspired by color temperature shifting tools and brightness
adaption software like:

- [redshift](https://wiki.archlinux.org/index.php/Redshift)
- [f.lux](https://justgetflux.com/news/pages/mac/)
- [Lumen](https://github.com/anishathalye/lumen)

---

### Example usage

##### Switching themes on time of day

Example usage featuring [hemera-themes](https://github.com/GuidoSchmidt/emacs-hemera-theme)
and [nyx-theme](https://github.com/GuidoSchmidt/emacs-nyx-theme) (with use-package). Make sure
to use `:defer` keyword. Omitting it may lead to broken colors
(see [issue 9](https://github.com/guidoschmidt/circadian.el/issues/9)):

```elisp
;; Install additinal themes from melpa
;; make sure to use :defer keyword
(use-package hemera-theme :ensure :defer)
(use-package nyx-theme :ensure :defer)

(use-package circadian
  :ensure t
  :config
  (setq circadian-themes '(("8:00" . hemera)
                           ("19:30" . nyx)))
  (circadian-setup))
```

##### Switching themes on sunrise & sunset

Be sure to set your latitude and longitude (Get them e.g. at [latlong.net](https://www.latlong.net/)):

```elisp
;; Install additinal themes from melpa
;; make sure to use :defer keyword
(use-package apropospriate-theme :ensure :defer)
(use-package nord-theme :ensure :defer)

(use-package circadian
  :ensure t
  :config
  (setq calendar-latitude 49.0)
  (setq calendar-longitude 8.5)
  (setq circadian-themes '((:sunrise . apropospriate-light)
                           (:sunset  . nord)))
  (circadian-setup))
```

---

### Hooks

**circadian** provides two hooks:

- `circadian-before-load-theme-hook`
- `circadian-after-load-theme-hook`

e.g. I like to override any themes cursor color to a very bright color via:

```elisp
(add-hook 'circadian-after-load-theme-hook
          #'(lambda (theme)
              ;; Line numbers appearance
              (setq linum-format 'linum-format-func)
              ;; Cursor
              (set-default 'cursor-type 'box)
              (set-cursor-color "#F52503")))
```

---

### Todos & Ideas

- Is it possible to interpolate between themes/colors?
- Can brightness sensors (e.g. laptops) be queried to control dimming?
- Load themes by mode [reddit.com/r/emacs](https://www.reddit.com/r/emacs/comments/72ukrx/theme_preferences/)
- Load themes by machine name [reddit.com/r/emacs](https://www.reddit.com/r/emacs/comments/72ukrx/theme_preferences/)
- Load themes by wifi/location? [reddit.com/r/emacs](https://www.reddit.com/r/emacs/comments/72ukrx/theme_preferences/)

---

### Development

Install Emacs [cask](https://github.com/cask/cask) environment. On macOS you
canr use [homebrew](https://brew.sh/) with: `brew install cask`.

0. Clone the circadian.el repository: `git clone git@github.com:guidoschmidt/circadian.el.git`
1. `cd` into the git project directory with `cd circadian.el` and install dependencies via [`cask`](https://github.com/cask/cask)
2. Run test using `make test`


CHANGELOG.md
# Change Log:

### 0.3.3
- Avoid redundant theme-loading: checking whether a desired theme is already
  loaded in `circadian-enable-theme`. `run-at-time` did fire when
  `circadian-setup` has already been overdue (e.g. during system hibernation). 
  Big kudos to @tangxinfa for the pull request.
- Fix #25: fixes wrong behavior when using time strings
- Fix #27: prompts error when (solar-sunrise-sunset (calendar-current-date))
  returns nil for requsted sunrise/sunset times

### 0.3.2
- Error messages in `circadian-match-sun` if sunset or sunrise could not be
  parsed due to time zone settings. Timezone issues should be checked/evaluated in future
  development and versions.

### 0.3.1
- Add theme loading hooks: `circadian-before-load-theme-hook` and
  `circadian-after-load-theme-hook`

### 0.3.0
- Implement configuration with keywords `:sunrise` and `:sunset`
- Rename `circadian-mapcar` to `circadian-mapc`
- Fixed time-comparison bug in
  in `circadian-compare-time-strings`
- Renamed `circadian-compare-time-strings` to more appropriate name:
  `circadian-a-earlier-b-p`

### 0.2.3
- Use -*- lexical-binding: t -*-
- Requiring cl-lib
- Prefixed cl function like `cl-first', `cl-remove-if'
- `mapcar' -> `mapc'
- Swapped argument order for `circadian-filter-inactivate-themes'
- Bugfix: load the last theme from `circadian-themes', when the first
time slot lies in the future

### 0.2.2
- Added testing (+ configuration for travis CI)
- Changed arguments of `circadian-filter-inactivate-themes' to accept
the current time string + themes asoc list due to gain testability
of that function with various time strings (see `tests/').

### 0.2.1
- Add function to load the latest overdue theme to `circadian-setup'

### 0.2.0
- nyx-theme and hemera-theme live in their own repos from now on:
nyx: https://github.com/GuidoSchmidt/emacs-nyx-theme
hemera: https://github.com/GuidoSchmidt/emacs-hemera-theme
- Use default themes for default configuration of `circadian-themes'
- Re-implemented configuration using associated list and timers
(thanks to Steve Purcell for pointing me into this direction)

### 0.1.0
- Initial release
- Variables for day/night hour
- Themes included: hemera-theme, nyx-theme


test.el
;;; circadian.el-test.el --- Tests for circadian.el
;;; Commentary:

;;; Code:
(require 'cl-lib)
(require 'el-mock)
(require 'benchmark)

(load (expand-file-name "circadian.el" default-directory))

(ert-deftest test-circadian-filter-and-activate-themes ()
  "Test filtering of `circadian-themes' list`."
  (setq circadian-themes '(("5:01"  . wombat)
                           ("14:47" . tango)
                           ("23:59" . adwaita)))
  (setq circadian-themes-parsed (circadian-themes-parse))

  ;; Before 5:01
  (let ((time-now '(4 10)))
    (should (equal 0 (length (circadian-filter-inactivate-themes
                              circadian-themes-parsed
                              time-now)))))
  (with-mock
   (stub circadian-now-time => '(5 0 0))
   (circadian-activate-latest-theme)
   (should (equal 'adwaita (cl-first custom-enabled-themes))))

  ;; After 5:01, before 14:47
  (let ((time-now '(5 2)))
    (should (equal 1 (length (circadian-filter-inactivate-themes
                              circadian-themes-parsed
                              time-now)))))
  (with-mock
   (stub circadian-now-time => '(5 2 0))
   (circadian-activate-latest-theme)
   (should (equal 'wombat (cl-first custom-enabled-themes))))

  ;; After 14:47, before 23:59
  (let ((time-now '(14 47)))
    (should (equal 2 (length (circadian-filter-inactivate-themes
                              circadian-themes-parsed
                              time-now)))))
  (with-mock
   (stub circadian-now-time => '(14 47 1))
   (circadian-activate-latest-theme)
   (should (equal 'tango (cl-first custom-enabled-themes))))

  ;; After 23:59
  (let ((time-now '(23 59)))
    (should (equal 3 (length (circadian-filter-inactivate-themes
                              circadian-themes-parsed
                              time-now)))))
  (with-mock
   (stub circadian-now-time => '(23 59 15))
   (circadian-activate-latest-theme)
   (should (equal 'adwaita (cl-first custom-enabled-themes))))

  ;; Surpassing midnight
  (let ((time-now '(0 2)))
    (should (equal 0 (length (circadian-filter-inactivate-themes
                              circadian-themes-parsed
                              time-now)))))
  (with-mock
   (stub circadian-now-time => '(0 2 10))
   (circadian-activate-latest-theme)
   (should (equal 'adwaita (cl-first custom-enabled-themes)))))



(ert-deftest test-circadian-activate-latest-theme ()
  "Test `circadian-activate-latest-theme' used in `circadian-setup'."
  ;; (print "→ TEST: circadian-activate-latest-theme")
  (setq circadian-themes '(("7:00" . wombat)
                           ("16:00" . tango)))
  (with-mock
   (stub circadian-now-time => '(7 21 0))
   (circadian-activate-latest-theme)
   (should (equal 'wombat (cl-first custom-enabled-themes))))
  (with-mock
   (stub circadian-now-time => '(17 0 0))
   (circadian-activate-latest-theme)
   (should (equal 'tango (cl-first custom-enabled-themes)))))



(ert-deftest test-circadian-sunrise-sunset ()
  "Test :sunrise and :sunset keywords for theme switching.
@TODO currently failing, needs a fix"
  (setq calendar-latitude 49.329896)
  (setq calendar-longitude 8.570925)
  (setq circadian-themes '((:sunrise . wombat)
                           (:sunset  . adwaita)))
  (circadian-setup)

  (with-mock
    (stub circadian-now-time => '(14 21 0))
    (circadian-activate-latest-theme)
    (should (equal 'wombat (cl-first custom-enabled-themes))))

  (with-mock
    (stub circadian-now-time-string => '(16 50 0))
    (circadian-activate-latest-theme)
    (should (equal 'adwaita (cl-first custom-enabled-themes)))))



(ert-deftest test-circadian-sunrise-sunset-timezones ()
  "Test :sunrise and :sunset for different time zones (TODO)."
  ;; (print "→ TEST: sunrise/sunset (New York Time Zone)")
  (setq calendar-latitude   40.712775)
  (setq calendar-longitude -74.005973)
  (setq calendar-time-zone -360)
  (setq circadian-themes '((:sunrise . wombat)
                           (:sunset . adwaita))))



(ert-deftest test-circadian-time-comparisons ()
  "Test comparison of time strings.
Time A: 17:59
Time B: 17:58
B should be earlier than A
=> `circadian-a-earlier-b-p' should return t."
  ;; (print "→ TEST: time comparisons")
  (should (equal t (circadian-a-earlier-b-p '(7 50) '(7 51))))
  (should (equal nil (circadian-a-earlier-b-p '(19 20) '(19 19))))
  (should (equal t (circadian-a-earlier-b-p '(20 20) '(20 20)))))



(ert-deftest test-circadian-setup-benchmark ()
  "Benchmark (circadian-setup)."
  (setq calendar-latitude 49)
  (setq calendar-longitude 5)
  (setq circadian-themes '((:sunrise . wombat)
                           (:sunset  . tango)))
  (let ((elapsed (benchmark-elapse (circadian-setup))))
    (should (equal t (< elapsed 0.05)))
    (print (concat "(circadian-setup) took " (format "%.10f seconds" elapsed)))))



(ert-deftest test-circadian-invalid-solar-sunrise-sunset ()
  :expected-result :failed
  "Test to address issue #27.
https://github.com/guidoschmidt/circadian.el/issues/27"
  (setq calendar-latitude 79.482623)
  (setq calendar-longitude 5.318703)
  (setq circadian-themes '((:sunrise . wombat)
                           (:sunset  . tango)))
  (circadian-setup))



(defvar test-order '(member
                     test-circadian-filter-and-activate-themes
                     test-circadian-activate-latest-theme
                     test-circadian-sunrise-sunset
                     test-circadian-sunrise-sunset-timezones
                     test-circadian-time-comparisons
                     test-circadian-setup-benchmark
                     test-circadian-invalid-solar-sunrise-sunset))

(provide 'circadian.el-test)
;;; circadian.el-test.el ends here


test/test-helper.el
;;; test-helper.el --- Helpers for circadian.el-test.el

;;; test-helper.el ends here


test/circadian.el-test.el
;;; circadian.el-test.el --- Tests for circadian.el
;;; Commentary:

;;; Code:
(require 'cl-lib)
(require 'el-mock)
(require 'benchmark)
(require 'circadian)

(ert-deftest test-circadian-filter-and-activate-themes ()
  "Test filtering of `circadian-themes' list`."
  (setq circadian-themes '(("5:01"  . wombat)
                           ("14:47" . tango)
                           ("23:59" . adwaita)))
  (setq circadian-themes-parsed (circadian-themes-parse))

  ;; Before 5:01
  (let ((time-now '(4 10)))
    (should (equal 0 (length (circadian-filter-inactivate-themes
                              circadian-themes-parsed
                              time-now)))))
  (with-mock
   (stub circadian-now-time => '(5 0 0))
   (circadian-activate-latest-theme)
   (should (equal 'adwaita (cl-first custom-enabled-themes))))

  ;; After 5:01, before 14:47
  (let ((time-now '(5 2)))
    (should (equal 1 (length (circadian-filter-inactivate-themes
                              circadian-themes-parsed
                              time-now)))))
  (with-mock
   (stub circadian-now-time => '(5 2 0))
   (circadian-activate-latest-theme)
   (should (equal 'wombat (cl-first custom-enabled-themes))))

  ;; After 14:47, before 23:59
  (let ((time-now '(14 47)))
    (should (equal 2 (length (circadian-filter-inactivate-themes
                              circadian-themes-parsed
                              time-now)))))
  (with-mock
   (stub circadian-now-time => '(14 47 1))
   (circadian-activate-latest-theme)
   (should (equal 'tango (cl-first custom-enabled-themes))))

  ;; After 23:59
  (let ((time-now '(23 59)))
    (should (equal 3 (length (circadian-filter-inactivate-themes
                              circadian-themes-parsed
                              time-now)))))
  (with-mock
   (stub circadian-now-time => '(23 59 15))
   (circadian-activate-latest-theme)
   (should (equal 'adwaita (cl-first custom-enabled-themes))))

  ;; Surpassing midnight
  (let ((time-now '(0 2)))
    (should (equal 0 (length (circadian-filter-inactivate-themes
                              circadian-themes-parsed
                              time-now)))))
  (with-mock
   (stub circadian-now-time => '(0 2 10))
   (circadian-activate-latest-theme)
   (should (equal 'adwaita (cl-first custom-enabled-themes)))))



(ert-deftest test-circadian-activate-latest-theme ()
  "Test `circadian-activate-latest-theme' used in `circadian-setup'."
  ;; (print "→ TEST: circadian-activate-latest-theme")
  (setq circadian-themes '(("7:00" . wombat)
                           ("16:00" . tango)))
  (with-mock
   (stub circadian-now-time => '(7 21 0))
   (circadian-activate-latest-theme)
   (should (equal 'wombat (cl-first custom-enabled-themes))))
  (with-mock
   (stub circadian-now-time => '(17 0 0))
   (circadian-activate-latest-theme)
   (should (equal 'tango (cl-first custom-enabled-themes)))))



(ert-deftest test-circadian-sunrise-sunset ()
  "Test :sunrise and :sunset keywords for theme switching.
@TODO currently failing, needs a fix"
  (setq calendar-latitude 49.329896)
  (setq calendar-longitude 8.570925)
  (setq circadian-themes '((:sunrise . wombat)
                           (:sunset  . adwaita)))
  (circadian-setup)

  (with-mock
    (print (circadian-sunrise))
    (stub circadian-now-time => '(14 21 0))
    (circadian-activate-latest-theme)
    (should (equal 'wombat (cl-first custom-enabled-themes))))

  (with-mock
    (print (circadian-sunset))
    (stub circadian-now-time-string => '(16 50 0))
    (circadian-activate-latest-theme)
    (should (equal 'adwaita (cl-first custom-enabled-themes)))))



(ert-deftest test-circadian-sunrise-sunset-timezones ()
  "Test :sunrise and :sunset for different time zones (TODO)."
  ;; (print "→ TEST: sunrise/sunset (New York Time Zone)")
  (setq calendar-latitude   40.712775)
  (setq calendar-longitude -74.005973)
  (setq calendar-time-zone -360)
  (setq circadian-themes '((:sunrise . wombat)
                           (:sunset . adwaita))))



(ert-deftest test-circadian-time-comparisons ()
  "Test comparison of time strings.
Time A: 17:59
Time B: 17:58
B should be earlier than A
=> `circadian-a-earlier-b-p' should return t."
  ;; (print "→ TEST: time comparisons")
  (should (equal t (circadian-a-earlier-b-p '(7 50) '(7 51))))
  (should (equal nil (circadian-a-earlier-b-p '(19 20) '(19 19))))
  (should (equal t (circadian-a-earlier-b-p '(20 20) '(20 20)))))



(ert-deftest test-circadian-setup-benchmark ()
  "Benchmark (circadian-setup)."
  (setq calendar-latitude 49)
  (setq calendar-longitude 5)
  (setq circadian-themes '((:sunrise . wombat)
                           (:sunset  . tango)))
  (let ((elapsed (benchmark-elapse (circadian-setup))))
    (should (equal t (< elapsed 0.01)))
    (print (concat "(circadian-setup) took " (format "%.10f seconds" elapsed)))))



(ert-deftest test-circadian-invalid-solar-sunrise-sunset ()
  :expected-result :failed
  "Test to address issue #27.
https://github.com/guidoschmidt/circadian.el/issues/27"
  (setq calendar-latitude 79.482623)
  (setq calendar-longitude 5.318703)
  (setq circadian-themes '((:sunrise . wombat)
                           (:sunset  . tango)))
  (circadian-setup))


(provide 'circadian.el-test)
;;; circadian.el-test.el ends here


circadian.el
;;; circadian.el --- Theme-switching based on daytime -*- lexical-binding: t -*-

;; Copyright (C) 2017 Guido Schmidt

;; Author: Guido Schmidt
;; Maintainer: Guido Schmidt <git@guidoschmidt.cc>
;; URL: https://github.com/GuidoSchmidt/circadian
;; Version: 0.3.3
;; Keywords: themes
;; Package-Requires: ((emacs "24.4"))

;; This file is part of GNU Emacs.

;; GNU Emacs is free software: you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; GNU Emacs is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with GNU Emacs.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:
;; Circadian provides automated theme switching based on daytime.
;;
;; Example usage (with `use-package') - make sure to properly set your
;; latitude & longitude:
;;
;; (use-package circadian
;;   :config
;;   (setq calendar-latitude 49.0)
;;   (setq calendar-longitude 8.5)
;;   (setq circadian-themes '((:sunrise . wombat)
;;                            ("8:00"   . tango)
;;                            (:sunset  . adwaita)
;;                            ("23:30"  . tango)))
;;   (circadian-setup))

;;; Code:
(require 'cl-lib)
(require 'solar)

(defcustom circadian-before-load-theme-hook nil
  "Functions to run before the theme is changed."
  :type 'hook
  :group 'circadian)

(defcustom circadian-after-load-theme-hook nil
  "Functions to run after the themes has changed."
  :type 'hook
  :group 'circadian)

(defcustom circadian-themes '(("7:30" . tango)
                              ("19:30" . wombat))
  "List of themes mapped to the time they should be loaded."
  :type 'alist
  :group 'circadian)

(defun circadian-enable-theme (theme)
  "Clear previous `custom-enabled-themes' and load THEME."
  (unless (equal (list theme) custom-enabled-themes)
    ;; Only load the argument theme, when `custom-enabled-themes'
    ;; does not contain it.
    (mapc #'disable-theme custom-enabled-themes)
    (condition-case nil
        (progn
          (run-hook-with-args 'circadian-before-load-theme-hook theme)
          (load-theme theme t)
          (let ((time (circadian-now-time)))
            (message "circadian.el → Enabled %s theme @ %02d:%02d:%02d"
                     theme (nth 0 time) (nth 1 time) (nth 2 time)))
          (run-hook-with-args 'circadian-after-load-theme-hook theme))
      (error "ERROR: circadian.el → Problem loading theme %s" theme))))

(defun circadian--encode-time (hour min)
  "Encode HOUR hours and MIN minutes into a valid format for `run-at-time'."
  (let ((now (decode-time)))
    (let ((day (nth 3 now))
          (month (nth 4 now))
          (year (nth 5 now))
          (zone (current-time-zone)))
      (encode-time 0 min hour day month year zone))))

(defun circadian-themes-parse ()
  "Parse `circadian-themes' and sort by time."
  (sort
   (mapcar
    (lambda (entry)
      (cons (circadian-match-sun (cl-first entry)) (cdr entry)))
    circadian-themes)
   (lambda (a b) (circadian-a-earlier-b-p (car a) (car b)))))

;;; --- TIME COMPARISONS
(defun circadian-now-time ()
  "Get the current time as string in the format (HH MM SS)."
  (reverse (cl-subseq (decode-time) 0 3)))

(defun circadian-a-earlier-b-p (time-a time-b)
  "Compare to time strings TIME-A and TIME-B by hour and minutes."
  (or (and (= (cl-first time-a) (cl-first time-b))
           (<= (cl-second time-a) (cl-second time-b)))
      (< (cl-first time-a) (cl-first time-b))))

(defun circadian-filter-inactivate-themes (theme-list now-time)
  "Filter THEME-LIST to consist of themes that are due NOW-TIME."
  (cl-remove-if (lambda (entry)
                  (let ((theme-time (cl-first entry)))
                    (not (circadian-a-earlier-b-p theme-time now-time))))
                theme-list))

(defun circadian-activate-latest-theme ()
  "Check which themes are overdue to be activated and load the last."
  (interactive)
  (let* ((themes (circadian-themes-parse))
         (now (circadian-now-time))
         (past-themes (circadian-filter-inactivate-themes themes now))
         (entry (car (last (or past-themes themes))))
         (theme (cdr entry))
         (next-entry (or (cadr (member entry themes))
                         (if (circadian-a-earlier-b-p (circadian-now-time) (cl-first entry))
                             (car themes))))
         (next-time (if next-entry
                        (circadian--encode-time
                         (cl-first (cl-first next-entry))
                         (cl-second (cl-first next-entry)))
                      (+ (* (+ (- 23 (cl-first now)) (cl-first (cl-first (cl-first themes)))) 60 60)
                         (* (+ (- 60 (cl-second now)) (cl-second (cl-first (cl-first themes)))) 60)))))
    (circadian-enable-theme theme)
    (cancel-function-timers #'circadian-activate-latest-theme)
    (run-at-time next-time nil #'circadian-activate-latest-theme)))

;; --- Sunset-sunrise
(defun circadian--frac-to-time (f)
  "Convert fractional time F to (HH MM)."
  (let ((l (cl-floor f)))
       (list (cl-first l)
             (floor (* 60 (cl-second l))))))

(defun circadian-sunrise ()
  "Get clean sunrise time string from Emacs' `sunset-sunrise'`."
  (let ((solar-result (solar-sunrise-sunset (calendar-current-date))))
    (let ((sunrise-numeric (cl-first (cl-first solar-result))))
      (if (equal nil sunrise-numeric)
          (error "No valid sunrise from solar-sunrise-sunset, consider using fixed time strings, e.g. (setq circadian-themes '((\"9:00\" . wombat) (\"20:00\" . tango)))")
        (circadian--frac-to-time sunrise-numeric)))))

(defun circadian-sunset ()
  "Get clean sunset time string from Emacs' `sunset-sunrise'`."
  (let ((solar-result (solar-sunrise-sunset (calendar-current-date))))
    (let ((sunset-numeric (cl-first (cl-second solar-result))))
      (if (equal nil sunset-numeric)
          (error "No valid sunset from solar-sunrise-sunset, consider using fixed time strings, e.g. (setq circadian-themes '((\"9:00\" . wombat) (\"20:00\" . tango)))")
        (circadian--frac-to-time sunset-numeric)))))

(defun circadian--string-to-time (input)
  "Parse INPUT string to `(HH MM)'."
  (cl-map 'list #'string-to-number (split-string input ":")))

(defun circadian-match-sun (input)
  "Match INPUT to a case for setting up timers."
  (cond ((cl-equalp input :sunrise)
         (let  ((sunrise (circadian-sunrise)))
           (if (equal sunrise "not")
               (error "Could not get valid sunset time — check your time zone settings"))
           (circadian-sunrise)))
        ((cl-equalp input :sunset)
         (let ((sunset (circadian-sunset)))
           (if (equal sunset "on")
               (error "Could not get valid sunset time — check your time zone settings"))
           (circadian-sunset)))
        ((stringp input) (circadian--string-to-time input))))

;;;###autoload
(defun circadian-setup ()
  "Setup circadian based on `circadian-themes'."
  (interactive)
  (circadian-activate-latest-theme))

(provide 'circadian)
;;; circadian.el ends here
